{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'nanoact.nanoact' from '/home/raingel/桌面/OOP_v2/nanoact/nanoact.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from nanoact import nanoact\n",
    "#Reload the module\n",
    "import importlib\n",
    "importlib.reload(nanoact)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dumb = nanoact.NanoAct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Clustering 578KIGS.fas\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bin/sh: 1: /nanoact/bin/mmseqs: not found\n",
      "/bin/sh: 1: /nanoact/bin/mmseqs: not found\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'./data/test/'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dumb.mmseqs_cluster(\"./data/temp/\",'./data/test/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster ./data/test/out_db ./data/test/clusterRes tmp -k 6 -s 7.5 --cluster-mode 2 \n",
      "\n",
      "MMseqs Version:                     \t7e2840992948ee89dcc336522dc98a74fe0adf00\n",
      "Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n",
      "Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                         \t7.5\n",
      "k-mer length                        \t6\n",
      "k-score                             \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                       \taa:21,nucl:5\n",
      "Max sequence length                 \t65535\n",
      "Max results per query               \t20\n",
      "Split database                      \t0\n",
      "Split mode                          \t2\n",
      "Split memory limit                  \t0\n",
      "Coverage threshold                  \t0.8\n",
      "Coverage mode                       \t0\n",
      "Compositional bias                  \t1\n",
      "Compositional bias                  \t1\n",
      "Diagonal scoring                    \ttrue\n",
      "Exact k-mer matching                \t0\n",
      "Mask residues                       \t1\n",
      "Mask residues probability           \t0.9\n",
      "Mask lower case residues            \t0\n",
      "Minimum diagonal score              \t15\n",
      "Selected taxa                       \t\n",
      "Include identical seq. id.          \tfalse\n",
      "Spaced k-mers                       \t1\n",
      "Preload mode                        \t0\n",
      "Pseudo count a                      \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                      \tsubstitution:4.100,context:5.800\n",
      "Spaced k-mer pattern                \t\n",
      "Local temporary path                \t\n",
      "Threads                             \t2\n",
      "Compressed                          \t0\n",
      "Verbosity                           \t3\n",
      "Add backtrace                       \tfalse\n",
      "Alignment mode                      \t3\n",
      "Alignment mode                      \t0\n",
      "Allow wrapped scoring               \tfalse\n",
      "E-value threshold                   \t0.001\n",
      "Seq. id. threshold                  \t0\n",
      "Min alignment length                \t0\n",
      "Seq. id. mode                       \t0\n",
      "Alternative alignments              \t0\n",
      "Max reject                          \t2147483647\n",
      "Max accept                          \t2147483647\n",
      "Score bias                          \t0\n",
      "Realign hits                        \tfalse\n",
      "Realign score bias                  \t-0.2\n",
      "Realign max seqs                    \t2147483647\n",
      "Correlation score weight            \t0\n",
      "Gap open cost                       \taa:11,nucl:5\n",
      "Gap extension cost                  \taa:1,nucl:2\n",
      "Zdrop                               \t40\n",
      "Rescore mode                        \t0\n",
      "Remove hits by seq. id. and coverage\tfalse\n",
      "Sort results                        \t0\n",
      "Cluster mode                        \t2\n",
      "Max connected component depth       \t1000\n",
      "Similarity type                     \t2\n",
      "Single step clustering              \tfalse\n",
      "Cascaded clustering steps           \t3\n",
      "Cluster reassign                    \tfalse\n",
      "Remove temporary files              \tfalse\n",
      "Force restart with latest tmp       \tfalse\n",
      "MPI runner                          \t\n",
      "k-mers per sequence                 \t21\n",
      "Scale k-mers per sequence           \taa:0.000,nucl:0.200\n",
      "Adjust k-mer length                 \tfalse\n",
      "Shift hash                          \t67\n",
      "Include only extendable             \tfalse\n",
      "Skip repeating k-mers               \tfalse\n",
      "\n",
      "Set cluster iterations to 3\n",
      "linclust ./data/test/out_db tmp/6739264664013116040/clu_redundancy tmp/6739264664013116040/linclust --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 2 --compressed 0 -v 3 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --alph-size aa:13,nucl:5 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 -k 0 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --rescore-mode 0 --filter-hits 0 --sort-results 0 --remove-tmp-files 0 --force-reuse 0 \n",
      "\n",
      "kmermatcher ./data/test/out_db tmp/6739264664013116040/linclust/13654149361498522562/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "kmermatcher ./data/test/out_db tmp/6739264664013116040/linclust/13654149361498522562/pref --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --alph-size aa:13,nucl:5 --min-seq-id 0 --kmer-per-seq 21 --spaced-kmer-mode 1 --kmer-per-seq-scale aa:0.000,nucl:0.200 --adjust-kmer-len 0 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --cov-mode 0 -k 0 -c 0.8 --max-seq-len 65535 --hash-shift 67 --split-memory-limit 0 --include-only-extendable 0 --ignore-multi-kmer 0 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Database size: 56 type: Aminoacid\n",
      "Reduced amino acid alphabet: (A S T) (C) (D B N) (E Q Z) (F Y) (G) (H) (I V) (K R) (L J M) (P) (W) (X) \n",
      "\n",
      "Generate k-mers list for 1 split\n",
      "[=================================================================] 56 0s 28ms\n",
      "Sort kmer 0h 0m 0s 3ms\n",
      "Sort by rep. sequence 0h 0m 0s 0ms\n",
      "Time for fill: 0h 0m 0s 0ms\n",
      "Time for merging to pref: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 141ms\n",
      "rescorediagonal ./data/test/out_db ./data/test/out_db tmp/6739264664013116040/linclust/13654149361498522562/pref tmp/6739264664013116040/linclust/13654149361498522562/pref_rescore1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 0 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0.5 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 56 0s 22ms\n",
      "Time for merging to pref_rescore1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 58ms\n",
      "clust ./data/test/out_db tmp/6739264664013116040/linclust/13654149361498522562/pref_rescore1 tmp/6739264664013116040/linclust/13654149361498522562/pre_clust --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy\n",
      "Total time: 0h 0m 0s 34ms\n",
      "\n",
      "Size of the sequence database: 56\n",
      "Size of the alignment database: 56\n",
      "Number of clusters: 4\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to pre_clust: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 42ms\n",
      "createsubdb tmp/6739264664013116040/linclust/13654149361498522562/order_redundancy ./data/test/out_db tmp/6739264664013116040/linclust/13654149361498522562/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "createsubdb tmp/6739264664013116040/linclust/13654149361498522562/order_redundancy tmp/6739264664013116040/linclust/13654149361498522562/pref tmp/6739264664013116040/linclust/13654149361498522562/pref_filter1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to pref_filter1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 12ms\n",
      "filterdb tmp/6739264664013116040/linclust/13654149361498522562/pref_filter1 tmp/6739264664013116040/linclust/13654149361498522562/pref_filter2 --filter-file tmp/6739264664013116040/linclust/13654149361498522562/order_redundancy --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 4 0s 18ms\n",
      "Time for merging to pref_filter2: 0h 0m 0s 5ms\n",
      "Time for processing: 0h 0m 0s 54ms\n",
      "rescorediagonal tmp/6739264664013116040/linclust/13654149361498522562/input_step_redundancy tmp/6739264664013116040/linclust/13654149361498522562/input_step_redundancy tmp/6739264664013116040/linclust/13654149361498522562/pref_filter2 tmp/6739264664013116040/linclust/13654149361498522562/pref_rescore2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 1 --wrapped-scoring 0 --filter-hits 1 -e 0.001 -c 0.8 -a 0 --cov-mode 0 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Can not find any score per column for coverage 0.800000 and sequence identity 0.000000. No hit will be filtered.\n",
      "[=================================================================] 4 0s 8ms\n",
      "Time for merging to pref_rescore2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 16ms\n",
      "align tmp/6739264664013116040/linclust/13654149361498522562/input_step_redundancy tmp/6739264664013116040/linclust/13654149361498522562/input_step_redundancy tmp/6739264664013116040/linclust/13654149361498522562/pref_rescore2 tmp/6739264664013116040/linclust/13654149361498522562/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4 type: Aminoacid\n",
      "Target database size: 4 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 4 0s 49ms\n",
      "Time for merging to aln: 0h 0m 0s 1ms\n",
      "6 alignments calculated\n",
      "4 sequence pairs passed the thresholds (0.666667 of overall calculated)\n",
      "1.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 71ms\n",
      "clust tmp/6739264664013116040/linclust/13654149361498522562/input_step_redundancy tmp/6739264664013116040/linclust/13654149361498522562/aln tmp/6739264664013116040/linclust/13654149361498522562/clust --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy\n",
      "Total time: 0h 0m 0s 36ms\n",
      "\n",
      "Size of the sequence database: 4\n",
      "Size of the alignment database: 4\n",
      "Number of clusters: 4\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clust: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 37ms\n",
      "mergeclusters ./data/test/out_db tmp/6739264664013116040/clu_redundancy tmp/6739264664013116040/linclust/13654149361498522562/pre_clust tmp/6739264664013116040/linclust/13654149361498522562/clust --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 4 0s 9ms\n",
      "Clustering step 2\n",
      "[=================================================================] 4 0s 17ms\n",
      "Write merged clustering\n",
      "[=================================================================] 56 0s 29ms\n",
      "Time for merging to clu_redundancy: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 36ms\n",
      "createsubdb tmp/6739264664013116040/clu_redundancy ./data/test/out_db tmp/6739264664013116040/input_step_redundancy -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step_redundancy: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "prefilter tmp/6739264664013116040/input_step_redundancy tmp/6739264664013116040/input_step_redundancy tmp/6739264664013116040/pref_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 1 -k 6 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 0 --comp-bias-corr-scale 1 --diag-score 0 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 0 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 4 type: Aminoacid\n",
      "Estimated memory consumption: 977M\n",
      "Target database size: 4 type: Aminoacid\n",
      "Index table k-mer threshold: 154 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 4 0s 16ms\n",
      "Index table: Masked residues: 2920\n",
      "Index table: fill\n",
      "[=================================================================] 4 0s 0ms\n",
      "Index statistics\n",
      "Entries:          75\n",
      "DB size:          488 MB\n",
      "Avg k-mer size:   0.000001\n",
      "Top 10 k-mers\n",
      "    ATGTAC\t3\n",
      "    TGCGCG\t3\n",
      "    ATTCGA\t2\n",
      "    GCGGAC\t2\n",
      "    CGAACC\t2\n",
      "    GCGACC\t2\n",
      "    GGTAGC\t2\n",
      "    TGAGTC\t2\n",
      "    TGGGGG\t2\n",
      "    GTGGTG\t2\n",
      "Time for index table init: 0h 0m 1s 432ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 154\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 4\n",
      "Target db start 1 to 4\n",
      "[=================================================================] 4 0s 1ms\n",
      "\n",
      "3.208902 k-mers per position\n",
      "81 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "2 sequences passed prefiltering per query sequence\n",
      "3 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 979ms\n",
      "align tmp/6739264664013116040/input_step_redundancy tmp/6739264664013116040/input_step_redundancy tmp/6739264664013116040/pref_step0 tmp/6739264664013116040/aln_step0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 0 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4 type: Aminoacid\n",
      "Target database size: 4 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 4 0s 7ms\n",
      "Time for merging to aln_step0: 0h 0m 0s 0ms\n",
      "11 alignments calculated\n",
      "4 sequence pairs passed the thresholds (0.363636 of overall calculated)\n",
      "1.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 33ms\n",
      "clust tmp/6739264664013116040/input_step_redundancy tmp/6739264664013116040/aln_step0 tmp/6739264664013116040/clu_step0 --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy\n",
      "Total time: 0h 0m 0s 0ms\n",
      "\n",
      "Size of the sequence database: 4\n",
      "Size of the alignment database: 4\n",
      "Number of clusters: 4\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step0: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "createsubdb tmp/6739264664013116040/clu_step0 tmp/6739264664013116040/input_step_redundancy tmp/6739264664013116040/input_step1 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter tmp/6739264664013116040/input_step1 tmp/6739264664013116040/input_step1 tmp/6739264664013116040/pref_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 4.25 -k 6 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 4 type: Aminoacid\n",
      "Estimated memory consumption: 977M\n",
      "Target database size: 4 type: Aminoacid\n",
      "Index table k-mer threshold: 125 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 4 0s 20ms\n",
      "Index table: Masked residues: 2920\n",
      "Index table: fill\n",
      "[=================================================================] 4 0s 0ms\n",
      "Index statistics\n",
      "Entries:          128\n",
      "DB size:          488 MB\n",
      "Avg k-mer size:   0.000002\n",
      "Top 10 k-mers\n",
      "    TAGCAA\t3\n",
      "    ATGTAC\t3\n",
      "    TGCGCG\t3\n",
      "    GAAAGA\t2\n",
      "    ATTCGA\t2\n",
      "    TTTGGA\t2\n",
      "    GCGGAC\t2\n",
      "    CGAACC\t2\n",
      "    GCGACC\t2\n",
      "    GGTAGC\t2\n",
      "Time for index table init: 0h 0m 0s 840ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 125\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 4\n",
      "Target db start 1 to 4\n",
      "[=================================================================] 4 0s 0ms\n",
      "\n",
      "0.013071 k-mers per position\n",
      "2 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "1 sequences passed prefiltering per query sequence\n",
      "1 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 318ms\n",
      "align tmp/6739264664013116040/input_step1 tmp/6739264664013116040/input_step1 tmp/6739264664013116040/pref_step1 tmp/6739264664013116040/aln_step1 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4 type: Aminoacid\n",
      "Target database size: 4 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 4 0s 11ms\n",
      "Time for merging to aln_step1: 0h 0m 0s 0ms\n",
      "4 alignments calculated\n",
      "4 sequence pairs passed the thresholds (1.000000 of overall calculated)\n",
      "1.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 13ms\n",
      "clust tmp/6739264664013116040/input_step1 tmp/6739264664013116040/aln_step1 tmp/6739264664013116040/clu_step1 --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy\n",
      "Total time: 0h 0m 0s 8ms\n",
      "\n",
      "Size of the sequence database: 4\n",
      "Size of the alignment database: 4\n",
      "Number of clusters: 4\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step1: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "createsubdb tmp/6739264664013116040/clu_step1 tmp/6739264664013116040/input_step1 tmp/6739264664013116040/input_step2 -v 3 --subdb-mode 1 \n",
      "\n",
      "Time for merging to input_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 0ms\n",
      "prefilter tmp/6739264664013116040/input_step2 tmp/6739264664013116040/input_step2 tmp/6739264664013116040/pref_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -s 7.5 -k 6 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 20 --split 0 --split-mode 2 --split-memory-limit 0 -c 0.8 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Query database size: 4 type: Aminoacid\n",
      "Estimated memory consumption: 977M\n",
      "Target database size: 4 type: Aminoacid\n",
      "Index table k-mer threshold: 96 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 4 0s 20ms\n",
      "Index table: Masked residues: 2920\n",
      "Index table: fill\n",
      "[=================================================================] 4 0s 0ms\n",
      "Index statistics\n",
      "Entries:          128\n",
      "DB size:          488 MB\n",
      "Avg k-mer size:   0.000002\n",
      "Top 10 k-mers\n",
      "    TAGCAA\t3\n",
      "    ATGTAC\t3\n",
      "    TGCGCG\t3\n",
      "    GAAAGA\t2\n",
      "    ATTCGA\t2\n",
      "    TTTGGA\t2\n",
      "    GCGGAC\t2\n",
      "    CGAACC\t2\n",
      "    GCGACC\t2\n",
      "    GGTAGC\t2\n",
      "Time for index table init: 0h 0m 0s 746ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 96\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 4\n",
      "Target db start 1 to 4\n",
      "[=================================================================] 4 0s 2ms\n",
      "\n",
      "0.913566 k-mers per position\n",
      "72 DB matches per sequence\n",
      "0 overflows\n",
      "0 queries produce too many hits (truncated result)\n",
      "2 sequences passed prefiltering per query sequence\n",
      "3 median result list length\n",
      "0 sequences with 0 size result lists\n",
      "Time for merging to pref_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 3s 892ms\n",
      "align tmp/6739264664013116040/input_step2 tmp/6739264664013116040/input_step2 tmp/6739264664013116040/pref_step2 tmp/6739264664013116040/aln_step2 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 0 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 0.001 --min-seq-id 0 --min-aln-len 0 --seq-id-mode 0 --alt-ali 0 -c 0.8 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 2147483647 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 4 type: Aminoacid\n",
      "Target database size: 4 type: Aminoacid\n",
      "Calculation of alignments\n",
      "[=================================================================] 4 0s 8ms\n",
      "Time for merging to aln_step2: 0h 0m 0s 0ms\n",
      "11 alignments calculated\n",
      "4 sequence pairs passed the thresholds (0.363636 of overall calculated)\n",
      "1.000000 hits per query sequence\n",
      "Time for processing: 0h 0m 0s 27ms\n",
      "clust tmp/6739264664013116040/input_step2 tmp/6739264664013116040/aln_step2 tmp/6739264664013116040/clu_step2 --cluster-mode 2 --max-iterations 1000 --similarity-type 2 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Clustering mode: Greedy\n",
      "Total time: 0h 0m 0s 0ms\n",
      "\n",
      "Size of the sequence database: 4\n",
      "Size of the alignment database: 4\n",
      "Number of clusters: 4\n",
      "\n",
      "Writing results 0h 0m 0s 0ms\n",
      "Time for merging to clu_step2: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 1ms\n",
      "mergeclusters ./data/test/out_db ./data/test/clusterRes tmp/6739264664013116040/clu_redundancy tmp/6739264664013116040/clu_step0 tmp/6739264664013116040/clu_step1 tmp/6739264664013116040/clu_step2 --threads 2 --compressed 0 -v 3 \n",
      "\n",
      "Clustering step 1\n",
      "[=================================================================] 4 0s 18ms\n",
      "Clustering step 2\n",
      "[=================================================================] 4 0s 26ms\n",
      "Clustering step 3\n",
      "[=================================================================] 4 0s 29ms\n",
      "Clustering step 4\n",
      "[=================================================================] 4 0s 29ms\n",
      "Write merged clustering\n",
      "[=================================================================] 56 0s 31ms\n",
      "Time for merging to clusterRes: 0h 0m 0s 1ms\n",
      "Time for processing: 0h 0m 0s 38ms\n",
      "\n",
      "createtsv ./data/test/out_db ./data/test/out_db ./data/test/clusterRes ./data/test/clusterRes.tsv \n",
      "\n",
      "MMseqs Version:                 \t7e2840992948ee89dcc336522dc98a74fe0adf00\n",
      "First sequence as representative\tfalse\n",
      "Target column                   \t1\n",
      "Add full header                 \tfalse\n",
      "Sequence source                 \t0\n",
      "Database output                 \tfalse\n",
      "Threads                         \t2\n",
      "Compressed                      \t0\n",
      "Verbosity                       \t3\n",
      "\n",
      "Time for merging to clusterRes.tsv: 0h 0m 0s 3ms\n",
      "Time for processing: 0h 0m 0s 45ms\n",
      "\n",
      "createseqfiledb ./data/test/out_db ./data/test/clusterRes ./data/test/clusterRes.seq \n",
      "\n",
      "MMseqs Version:\t7e2840992948ee89dcc336522dc98a74fe0adf00\n",
      "Min sequences\t1\n",
      "Max sequences\t2147483647\n",
      "HH format    \tfalse\n",
      "Preload mode \t0\n",
      "Threads      \t2\n",
      "Compressed   \t0\n",
      "Verbosity    \t3\n",
      "\n",
      "[=================================================================] 4 0s 1ms\n",
      "Time for merging to clusterRes.seq: 0h 0m 0s 0ms\n",
      "Time for processing: 0h 0m 0s 21ms\n",
      "\n",
      "result2flat ./data/test/out_db ./data/test/out_db ./data/test/clusterRes.seq ./data/test/clusterRes.fas \n",
      "\n",
      "MMseqs Version: \t7e2840992948ee89dcc336522dc98a74fe0adf00\n",
      "Use fasta header\tfalse\n",
      "Verbosity       \t3\n",
      "\n",
      "Time for processing: 0h 0m 0s 14ms\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Convert fasta to DB\n",
    "name = \"578KIGS\"\n",
    "dumb._exec(f\"./nanoact/bin/mmseqs createdb data/4_deHead__dumb/{name}.fas ./data/test/out_db\")\n",
    "#--single-step-clustering Starts the single-step instead of the cascaded clustering workflow.\n",
    "#-s [float] Target sensitivity in the range  [1:7.5] (default= depended on the sequence identity).\n",
    "#--min-seq-id [float] list matches above this sequence identity [0.0:1.0] (default=0.0).\n",
    "#--cov-mode [int] \"0: coverage of query and target, 1: coverage of target [0:1] (default=0). -c [float] \"list matches above this fraction of covered residues (see cov-mode) [0.0:1.0] (default=0.8).\n",
    "#prefiltering data structure allocates alphabetSize^kmerSize * 8 byte (4^15*8=8GB).\n",
    "#Set kmerSize to 10 to reduce memory usage.\n",
    "print(dumb._exec(\"./nanoact/bin/mmseqs cluster  ./data/test/out_db ./data/test/clusterRes tmp -k 6 -s 7.5 --cluster-mode 2\")[0])\n",
    "print(dumb._exec(\"./nanoact/bin/mmseqs createtsv ./data/test/out_db ./data/test/out_db ./data/test/clusterRes ./data/test/clusterRes.tsv\")[0])\n",
    "print(dumb._exec(\"./nanoact/bin/mmseqs createseqfiledb  ./data/test/out_db ./data/test/clusterRes ./data/test/clusterRes.seq\")[0])\n",
    "print(dumb._exec(\"./nanoact/bin/mmseqs result2flat  ./data/test/out_db ./data/test/out_db ./data/test/clusterRes.seq ./data/test/clusterRes.fas\")[0])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
