{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[17:51:10] Temp folder is set to ./temp/, you can change it by NanoAct(TEMP='your_temp_folder')\n",
      "[17:51:10] We recommend not to set the temp folder to nfs, ftp, samba, google drive, etc. as it may cause unexpected errors\n"
     ]
    }
   ],
   "source": [
    "#載入並初始化nanoACT\n",
    "from nanoact import nanoact\n",
    "dumb = nanoact.NanoAct()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[15:35:02] Found fastq file: FAW73888_d763411f_40013a17_1.fast5.fastq.gz\n",
      "[15:35:02] Found fastq file: FAW73888_d763411f_40013a17_2.fast5.fastq.gz\n",
      "[15:35:02] Found fastq file: FAW73888_d763411f_40013a17_0.fast5.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "#將fastq.gz檔案合併成一個fastq檔案\n",
    "all_fastq = dumb.combine_fastq(src = \"./nanoact/sample/1_raw_reads/\",\n",
    "                               des = \"./nanoact/sample/2_merged_fastq/\",\n",
    "                               name = \"all.fastq\"\n",
    "                               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[15:35:24] Start Qualityfilt...\n",
      "[15:35:24] QSCORE: 7, MIN_LEN: 2500, MAX_LEN: 3500\n",
      "[15:35:30] 2481/12000 (20.67%) reads were passed quality filter\n"
     ]
    }
   ],
   "source": [
    "# 透過品質及長度過濾fastq檔案\n",
    "# See also nanofilt()\n",
    "filtered_fastq = dumb.qualityfilt(src = \"./nanoact/sample/2_merged_fastq/all.fastq\",\n",
    "                           des = \"./nanoact/sample/3_filtered_fastq/\",\n",
    "                           name = 'all.fastq',\n",
    "                           QSCORE = 7,  #recommended 7-9\n",
    "                           MIN_LEN = 2500, #depends on the length of your reads\n",
    "                           MAX_LEN = 3500 #depends on the length of your reads\n",
    "                           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[15:36:49] BARCODE_INDEX_FILE loaded\n",
      "[15:36:51] 498/2481 (20.07%) reads were demultiplexed successfully\n"
     ]
    }
   ],
   "source": [
    "##透過barcode_index將fastq檔案分類(demultiplex)\n",
    "#modified from Krehenwinkel, Henrik, Aaron Pomerantz, James B. Henderson, Susan R. Kennedy, Jun Ying Lim, Varun Swamy, Juan Diego Shoobridge, et al. 2019. “Nanopore Sequencing of Long Ribosomal DNA Amplicons Enables Portable and Simple Biodiversity Assessments with High Phylogenetic Resolution across Broad Taxonomic Scale.” GigaScience 8 (5). https://doi.org/10.1093/gigascience/giz006.\n",
    "demultiplexed = dumb.singlebar(src = \"./nanoact/sample/3_filtered_fastq/all.fastq\", #Input: 單個fastq檔案，例如 all.fastq\n",
    "                               des =  \"./nanoact/sample/4_demultiplex/\", #Output: 一個資料夾，程式會在該資料夾中輸出以SampleID為檔名的fastq檔案或是fasta檔案（由output_format決定），例如 SampleID.fastq\n",
    "                               BARCODE_INDEX_FILE =  \"./nanoact/sample/barcode.csv\", #BARCODE_INDEX_FILE: barcode資料庫，可以是csv或是tsv檔案，例如 barcode.csv。必須包含SampleID, FwIndex, RvAnchor、ExpectedLength四個欄位。\n",
    "                               mismatch_ratio_f = 0.1, #mismatch_ratio_f: FwIndex容許的錯誤率，預設為0.15。例如barcode長度為20bp，則容許0.15*20=3bp的錯誤(edit distance)。\n",
    "                               mismatch_ratio_r = 0.1, #mismatch_ratio_r: RvAnchor容許的錯誤率，預設為0.15。\n",
    "                               expected_length_variation = 0.3, #expected_length_variation: 預期的read長度變異，預設為0.3。例如ExpectedLengt為300bp，則容許0.3*300=90bp的變異。\n",
    "                               search_range=150, #search_range: 搜尋barcode的範圍，預設為150bp。代表搜尋範圍為前150bp和後150bp。\n",
    "                               output_format=\"both\" #output_format: 輸出檔案的格式，預設為both。可以是fastq或是fasta或是fas(fasta格式但附檔名為fas)。both代表同時輸出fastq和fasta。\n",
    "                               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[15:37:12] T24.fas is not in the accepted input format, skipping\n",
      "[15:37:12] T4.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T2.fastq\n",
      "[15:37:12] T2.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T6.fastq\n",
      "[15:37:12] T15.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T4.fastq\n",
      "[15:37:12] T13.fas is not in the accepted input format, skipping\n",
      "[15:37:12] T5.fas is not in the accepted input format, skipping\n",
      "[15:37:12] T27.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T14.fastq\n",
      "[15:37:12] T17.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T20.fastq\n",
      "[15:37:12] T8.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T18.fastq\n",
      "[15:37:12] Processing T3.fastq\n",
      "[15:37:12] T11.fas is not in the accepted input format, skipping\n",
      "[15:37:12] T19.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T25.fastq\n",
      "[15:37:12] T28.fas is not in the accepted input format, skipping\n",
      "[15:37:12] T21.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T7.fastq\n",
      "[15:37:12] T18.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T24.fastq\n",
      "[15:37:12] T25.fas is not in the accepted input format, skipping\n",
      "[15:37:12] Processing T27.fastq\n",
      "[15:37:12] Processing T29.fastq\n",
      "[15:37:12] Processing T11.fastq\n",
      "[15:37:13] T7.fas is not in the accepted input format, skipping\n",
      "[15:37:13] Processing T21.fastq\n",
      "[15:37:13] Processing T30.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[15:37:13] T1.fas is not in the accepted input format, skipping\n",
      "[15:37:13] Processing T31.fastq\n",
      "[15:37:13] T22.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T30.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T23.fas is not in the accepted input format, skipping\n",
      "[15:37:13] Processing T5.fastq\n",
      "[15:37:13] Processing T28.fastq\n",
      "[15:37:13] Processing T13.fastq\n",
      "[15:37:13] Processing T12.fastq\n",
      "[15:37:13] Processing T1.fastq\n",
      "[15:37:13] T6.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T3.fas is not in the accepted input format, skipping\n",
      "[15:37:13] Processing T9.fastq\n",
      "[15:37:13] Processing T23.fastq\n",
      "[15:37:13] T12.fas is not in the accepted input format, skipping\n",
      "[15:37:13] 2_Singlebar_stat.csv is not in the accepted input format, skipping\n",
      "[15:37:13] Processing T16.fastq\n",
      "[15:37:13] Processing T19.fastq\n",
      "[15:37:13] T9.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T20.fas is not in the accepted input format, skipping\n",
      "[15:37:13] Processing T22.fastq\n",
      "[15:37:13] Processing T15.fastq\n",
      "[15:37:13] Processing T10.fastq\n",
      "[15:37:13] Processing T17.fastq\n",
      "[15:37:13] Processing T8.fastq\n",
      "[15:37:13] T29.fas is not in the accepted input format, skipping\n",
      "[15:37:13] Processing T26.fastq\n",
      "[15:37:13] T31.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T14.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T26.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T10.fas is not in the accepted input format, skipping\n",
      "[15:37:13] T16.fas is not in the accepted input format, skipping\n"
     ]
    }
   ],
   "source": [
    "##Orientation，透過FwPrimer和RvPrimer將反向的序列轉換成正向的序列\n",
    "orientation = dumb.orientation( src= \"./nanoact/sample/4_demultiplex/\",  #Input: 一個資料夾，資料夾中包含以SampleID為檔名的fastq檔案或是fasta檔案（由input_format決定），例如 SampleID.fastq\n",
    "                                des= \"./nanoact/sample/5_orientation/\",  #Output: 一個資料夾，程式會在該資料夾中輸出以SampleID為檔名的fastq檔案或是fasta檔案（由output_format決定），例如 SampleID.fastq\n",
    "                                input_format = \"fastq\", #input_format: 輸入檔案的格式，預設為fastq。可以是fastq或是fasta或是fas(副檔名為fas之fasta格式檔案)。\n",
    "                                output_format = \"both\", #output_format: 輸出檔案的格式，預設為both。可以是fastq或是fasta。both代表同時輸出fastq和fasta。\n",
    "                                BARCODE_INDEX_FILE = \"./nanoact/sample/barcode.csv\", #BARCODE_INDEX_FILE: barcode資料庫，可以是csv或是tsv檔案，例如 barcode.csv。\n",
    "                                FwPrimer = \"FwPrimer\", #FwPrimer: FwPrimer的欄位名稱，預設為FwPrimer。\n",
    "                                RvPrimer = \"RvPrimer\", #RvPrimer: RvPrimer的欄位名稱，預設為RvPrimer。\n",
    "                                search_range=200, #search_range: 搜尋FwPrimer和RvPrimer的範圍，預設為200bp。代表搜尋範圍為前200bp和後200bp。\n",
    "                                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notice: mode is set to 'case', arguments other than src, des, fw_offset, rv_offset,input_format, output_format will be ignored\n",
      "[15:37:20] T28.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T24.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T4.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T2.fastq\n",
      "[15:37:20] T2.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T6.fastq\n",
      "[15:37:20] T15.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T4.fastq\n",
      "[15:37:20] T4.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T13.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T5.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T22.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T27.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T14.fastq\n",
      "[15:37:20] T15.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T17.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T8.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T1.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T20.fastq\n",
      "[15:37:20] T8.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T18.fastq\n",
      "[15:37:20] Tirmming T3.fastq\n",
      "[15:37:20] T11.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T19.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T25.fastq\n",
      "[15:37:20] T29.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T28.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T21.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T7.fastq\n",
      "[15:37:20] T24.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T30.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T2.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T18.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T24.fastq\n",
      "[15:37:20] T25.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T16.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T27.fastq\n",
      "[15:37:20] T11.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T12.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T29.fastq\n",
      "[15:37:20] T6.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T11.fastq\n",
      "[15:37:20] T7.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T21.fastq\n",
      "[15:37:20] Tirmming T30.fastq\n",
      "[15:37:20] T17.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T1.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T27.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T19.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T31.fastq\n",
      "[15:37:20] T22.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T30.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T23.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T5.fastq\n",
      "[15:37:20] T5.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T14.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T18.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T28.fastq\n",
      "[15:37:20] T7.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T13.fastq\n",
      "[15:37:20] T25.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T10.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T13.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T12.fastq\n",
      "[15:37:20] Tirmming T1.fastq\n",
      "[15:37:20] T6.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T23.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T20.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T3.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T9.fastq\n",
      "[15:37:20] Tirmming T23.fastq\n",
      "[15:37:20] T12.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T16.fastq\n",
      "[15:37:20] T21.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T19.fastq\n",
      "[15:37:20] T31.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T9.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T20.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T22.fastq\n",
      "[15:37:20] Tirmming T15.fastq\n",
      "[15:37:20] T26.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T10.fastq\n",
      "[15:37:20] Tirmming T17.fastq\n",
      "[15:37:20] Tirmming T8.fastq\n",
      "[15:37:20] T29.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] Tirmming T26.fastq\n",
      "[15:37:20] T31.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T9.fasta is not in the accepted inpute format, skipping\n",
      "[15:37:20] T14.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T26.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T10.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T16.fas is not in the accepted inpute format, skipping\n",
      "[15:37:20] T3.fasta is not in the accepted inpute format, skipping\n"
     ]
    }
   ],
   "source": [
    "# 將barcode, Primer等人造序列去除\n",
    "# 有兩種模式，一種是table，一種是case\n",
    "# model = table時，需要輸入barcode_index_file，並且需要指定fw_col和rv_col。程式會透過barcode_index_file中的SampleID, FwPrimer, RvPrimer欄位來去除人造序列。\n",
    "# model = case時，則是利用singlebar()分類時，會對raw reads找到的FwIndex及RvAnchor做小寫標記，trim_reads()依據小寫標記的序列位置來去除人造序列。\n",
    "\n",
    "dumb.trim_reads (src=\"./nanoact/sample/5_orientation/\",   #Input 一個資料夾，資料夾中包含以SampleID為檔名的fastq檔案或是fasta檔案（由input_format決定），例如 SampleID.fastq\n",
    "                 des=\"./nanoact/sample/6_trimmed/\",   #Output 一個資料夾，程式會在該資料夾中輸出以SampleID為檔名的fastq檔案或是fasta檔案（由output_format決定），例如 SampleID.fastq\n",
    "                BARCODE_INDEX_FILE = \"./nanoact/sample/barcode.csv\", #barcode資料庫，可以是csv或是tsv檔案，例如 barcode.csv。僅在mode為table時需要。\n",
    "                fw_col = \"FwIndex\",   #forward的欄位名稱，預設為FwPrimer。僅在mode為table時需要。\n",
    "                rv_col = \"RvPrimer\",  #reverse的欄位名稱，預設為RvPrimer。僅在mode為table時需要。\n",
    "                input_format=\"fastq\", #input_format: 輸入檔案的格式，預設為fastq。可以是fastq或是fasta。\n",
    "                output_format=\"both\", #output_format: 輸出檔案的格式，預設為both。可以是fastq或是fasta。both代表同時輸出fastq和fasta。\n",
    "                mode=\"case\", #mode: 輸入barcode的模式，預設為table。可以是table或是case。table代表使用barcode_index_file，case代表使用singlebar的小寫標記的barcode。\n",
    "                fw_offset = 0, #從距離找到的切除位點開始往後切除幾個bp，預設為0，可以是負數。例如fw_offset=-10，則從距離找到的切除位點開始往前切除10個bp。\n",
    "                rv_offset = 0, #從距離找到的切除位點開始往前切除幾個bp，預設為0，可以是負數。例如rv_offset=-10，則從距離找到的切除位點開始往後切除10個bp。\n",
    "                mismatch_ratio_f = 0.1, #FwIndex容許的錯誤率，預設為0.15。例如barcode長度為20bp，則容許0.15*20=3bp的錯誤(edit distance)。 僅在mode為table時需要。\n",
    "                mismatch_ratio_r = 0.1, #RvAnchor容許的錯誤率，預設為0.15。 僅在mode為table時需要。\n",
    "                discard_no_match = False, #如果沒有找到barcode，是否捨棄該序列，預設為False。 僅在mode為table時需要。\n",
    "                check_both_directions = True, #是否同時檢查正向和反向序列，如果反向序列中找到barcode，則輸出反向序列，預設為True。 僅在mode為table時需要。\n",
    "                reverse_complement_rv_col = True, #是否將RvAnchor欄位的序列反向互補後再進行搜尋，預設為True。 僅在mode為table時需要。\n",
    "                search_range=100, #搜尋FwIndex和RvAnchor的範圍，預設為200bp。代表搜尋範圍為前200bp和後200bp。 僅在mode為table時需要。\n",
    "                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[15:37:27] T24.fas is not in the accepted input format, skipping\n",
      "[15:37:27] T4.fas is not in the accepted input format, skipping\n",
      "[15:37:27] Clustering T2.fastq\n",
      "[15:37:32] Number of clusters 5\n",
      "[15:37:32] T2.fas is not in the accepted input format, skipping\n",
      "[15:37:32] Clustering T6.fastq\n",
      "[15:37:36] Number of clusters 1\n",
      "[15:37:36] T15.fas is not in the accepted input format, skipping\n",
      "[15:37:36] Clustering T4.fastq\n",
      "[15:37:41] Number of clusters 3\n",
      "[15:37:41] T13.fas is not in the accepted input format, skipping\n",
      "[15:37:41] T5.fas is not in the accepted input format, skipping\n",
      "[15:37:41] T27.fas is not in the accepted input format, skipping\n",
      "[15:37:41] Clustering T14.fastq\n",
      "[15:37:45] Number of clusters 5\n",
      "[15:37:45] T17.fas is not in the accepted input format, skipping\n",
      "[15:37:45] Clustering T20.fastq\n",
      "[15:37:48] Number of clusters 2\n",
      "[15:37:48] T8.fas is not in the accepted input format, skipping\n",
      "[15:37:48] Clustering T18.fastq\n",
      "[15:37:52] Number of clusters 4\n",
      "[15:37:52] Clustering T3.fastq\n",
      "[15:37:56] Number of clusters 3\n",
      "[15:37:56] T11.fas is not in the accepted input format, skipping\n",
      "[15:37:56] T19.fas is not in the accepted input format, skipping\n",
      "[15:37:56] Clustering T25.fastq\n",
      "[15:37:59] Number of clusters 2\n",
      "[15:37:59] T28.fas is not in the accepted input format, skipping\n",
      "[15:37:59] T21.fas is not in the accepted input format, skipping\n",
      "[15:37:59] Clustering T7.fastq\n",
      "[15:38:02] Number of clusters 1\n",
      "[15:38:02] T18.fas is not in the accepted input format, skipping\n",
      "[15:38:02] Clustering T24.fastq\n",
      "[15:38:05] Number of clusters 2\n",
      "[15:38:05] T25.fas is not in the accepted input format, skipping\n",
      "[15:38:05] Clustering T27.fastq\n",
      "[15:38:09] Number of clusters 1\n",
      "[15:38:09] Clustering T29.fastq\n",
      "[15:38:12] Number of clusters 1\n",
      "[15:38:12] Clustering T11.fastq\n",
      "[15:38:16] Number of clusters 7\n",
      "[15:38:16] T7.fas is not in the accepted input format, skipping\n",
      "[15:38:16] Clustering T21.fastq\n",
      "[15:38:19] Number of clusters 2\n",
      "[15:38:20] Clustering T30.fastq\n",
      "[15:38:23] Number of clusters 6\n",
      "[15:38:23] T1.fas is not in the accepted input format, skipping\n",
      "[15:38:23] Clustering T31.fastq\n",
      "[15:38:26] Number of clusters 2\n",
      "[15:38:26] T22.fas is not in the accepted input format, skipping\n",
      "[15:38:26] T30.fas is not in the accepted input format, skipping\n",
      "[15:38:26] T23.fas is not in the accepted input format, skipping\n",
      "[15:38:26] Clustering T5.fastq\n",
      "[15:38:29] Number of clusters 1\n",
      "[15:38:29] Clustering T28.fastq\n",
      "[15:38:29] Error reading output file, [Errno 2] No such file or directory: './temp//cluster.fas'\n",
      "[15:38:29] Clustering T13.fastq\n",
      "[15:38:33] Number of clusters 3\n",
      "[15:38:33] Clustering T12.fastq\n",
      "[15:38:36] Number of clusters 2\n",
      "[15:38:36] Clustering T1.fastq\n",
      "[15:38:40] Number of clusters 1\n",
      "[15:38:40] T6.fas is not in the accepted input format, skipping\n",
      "[15:38:40] T3.fas is not in the accepted input format, skipping\n",
      "[15:38:40] Clustering T9.fastq\n",
      "[15:38:43] Number of clusters 7\n",
      "[15:38:43] Clustering T23.fastq\n",
      "[15:38:43] Error reading output file, [Errno 2] No such file or directory: './temp//cluster.fas'\n",
      "[15:38:43] T12.fas is not in the accepted input format, skipping\n",
      "[15:38:43] Clustering T16.fastq\n",
      "[15:38:44] Error reading output file, [Errno 2] No such file or directory: './temp//cluster.fas'\n",
      "[15:38:44] Clustering T19.fastq\n",
      "[15:38:47] Number of clusters 1\n",
      "[15:38:47] T9.fas is not in the accepted input format, skipping\n",
      "[15:38:47] T20.fas is not in the accepted input format, skipping\n",
      "[15:38:47] Clustering T22.fastq\n",
      "[15:38:47] Error reading output file, [Errno 2] No such file or directory: './temp//cluster.fas'\n",
      "[15:38:47] Clustering T15.fastq\n",
      "[15:38:47] Error reading output file, [Errno 2] No such file or directory: './temp//cluster.fas'\n",
      "[15:38:47] Clustering T10.fastq\n",
      "[15:38:50] Number of clusters 3\n",
      "[15:38:50] Clustering T17.fastq\n",
      "[15:38:54] Number of clusters 2\n",
      "[15:38:54] Clustering T8.fastq\n",
      "[15:38:57] Number of clusters 1\n",
      "[15:38:57] T29.fas is not in the accepted input format, skipping\n",
      "[15:38:57] Clustering T26.fastq\n",
      "[15:39:00] Number of clusters 1\n",
      "[15:39:00] T31.fas is not in the accepted input format, skipping\n",
      "[15:39:00] T14.fas is not in the accepted input format, skipping\n",
      "[15:39:00] T26.fas is not in the accepted input format, skipping\n",
      "[15:39:00] T10.fas is not in the accepted input format, skipping\n",
      "[15:39:00] T16.fas is not in the accepted input format, skipping\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'./nanoact/sample/7_cluster/'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Clustering，將序列分群。\n",
    "#一個barcode下，在許多情境下raw reads內可能包含許多異質的序列。\n",
    "#例如由土壤放大的16S rRNA gene，可能包含多種細菌的16S rRNA gene(meta-barcoding)。\n",
    "#又或是來自分析過程中的污染、非專一性放大產物、chimera等等。\n",
    "#若直接將異質的raw reads進行後續分析，會導致具有許多雜訊的結果\n",
    "#這裡介紹mmseqs，本工具也實作了其他的分群工具，請見nanoACT tutorial\n",
    "dumb.mmseqs_cluster(src = \"./nanoact/sample/6_trimmed/\",\n",
    "                    des = \"./nanoact/sample/7_cluster/\",\n",
    "                    min_seq_id=0.8, #min_seq_id sets the minimum sequence identity threshold for prefiltering (pre-clustering). \n",
    "                    #Sequences with pairwise similarity below this value will not be clustered together.\n",
    "                    #Set higher values for more stringent prefiltering. (Save time)\n",
    "                    cluster_mode=0, #See mmseqs2 document: https://mmseqs.com/latest/userguide.pdf\n",
    "                    cov_mode=0, #See mmseqs2 document: https://mmseqs.com/latest/userguide.pdf\n",
    "                    k=14, #k specifies the length of k-mers length that will be used to compute sequence similarity.\n",
    "                    #A k-mer is a substring of length k found in a sequence.\n",
    "                    #   Increasing the value of k increases the specificity of sequence similarity, but also increases\n",
    "                    #   the computational (memory) cost.\n",
    "                    #   If you receive a memory error, try reducing the value of k.\n",
    "                    kmer_per_seq = 20, #Only used when cluster_mode=\"linclust\", see mmseqs2 document: https://mmseqs.com/latest/userguide.pdf\n",
    "                    s=7.5,# s sets the sensitivity of the clustering. values between 1 and 8.5.\n",
    "                    #   Increasing the value of s increases the sensitivity of the clustering (more clusters),\n",
    "                    #   but also increases the false positive rate (more spurious clusters).\n",
    "                    #   If you receive too many clusters, try reducing the value of s.\n",
    "                    min_read_num=0, #min_read_num: This sets the minimum number of reads in a cluster.\n",
    "                    #    Clusters with fewer reads than this value will be discarded.\n",
    "                    input_format = \"fastq\", # input_format: 輸入檔案的格式，預設為fastq。可以是fastq或是fasta。\n",
    "                    output_format = \"both\", # output_format: 輸出檔案的格式，預設為both。可以是fastq或是fasta。both代表同時輸出fastq和fasta。\n",
    "                    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16:00:56] Working on T10_cluster_0_r7 ...\n",
      "[16:00:57] T11_cluster_4_r1.fas is not in the accepted input format, skipping\n",
      "[16:00:57] Working on T3_cluster_1_r1 ...\n",
      "[16:00:58] T13_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:00:58] Working on T2_cluster_4_r7 ...\n",
      "[16:01:00] Working on T20_cluster_1_r7 ...\n",
      "[16:01:02] Working on T9_cluster_0_r1 ...\n",
      "[16:01:02] T31_cluster_1_r2.fas is not in the accepted input format, skipping\n",
      "[16:01:02] T25_cluster_0_r5.fas is not in the accepted input format, skipping\n",
      "[16:01:02] Working on T3_cluster_2_r1 ...\n",
      "[16:01:03] T8_cluster_0_r2.fas is not in the accepted input format, skipping\n",
      "[16:01:03] T11_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:03] Working on T4_cluster_0_r1 ...\n",
      "[16:01:03] Working on T24_cluster_0_r1 ...\n",
      "[16:01:04] Working on T31_cluster_0_r1 ...\n",
      "[16:01:04] T25_cluster_1_r2.fas is not in the accepted input format, skipping\n",
      "[16:01:04] T1_cluster_0_r8.fas is not in the accepted input format, skipping\n",
      "[16:01:04] Working on T1_cluster_0_r8 ...\n",
      "[16:01:06] T29_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:06] Working on T12_cluster_0_r5 ...\n",
      "[16:01:07] T14_cluster_2_r17.fas is not in the accepted input format, skipping\n",
      "[16:01:07] T31_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:07] Working on T21_cluster_0_r1 ...\n",
      "[16:01:07] Working on T29_cluster_0_r1 ...\n",
      "[16:01:08] Working on T30_cluster_0_r1 ...\n",
      "[16:01:08] Working on T2_cluster_2_r1 ...\n",
      "[16:01:08] Working on T25_cluster_0_r5 ...\n",
      "[16:01:10] T11_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:10] Working on T18_cluster_2_r1 ...\n",
      "[16:01:10] T13_cluster_0_r103.fas is not in the accepted input format, skipping\n",
      "[16:01:10] Working on T11_cluster_3_r1 ...\n",
      "[16:01:10] T30_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:10] T4_cluster_1_r11.fas is not in the accepted input format, skipping\n",
      "[16:01:10] Working on T11_cluster_1_r1 ...\n",
      "[16:01:11] T19_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:11] Working on T20_cluster_0_r2 ...\n",
      "[16:01:11] T9_cluster_3_r15.fas is not in the accepted input format, skipping\n",
      "[16:01:11] Working on T30_cluster_2_r1 ...\n",
      "[16:01:12] Working on T14_cluster_2_r17 ...\n",
      "[16:01:15] T24_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:15] Working on T9_cluster_1_r1 ...\n",
      "[16:01:15] T4_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:15] Working on T25_cluster_1_r2 ...\n",
      "[16:01:16] T20_cluster_1_r7.fas is not in the accepted input format, skipping\n",
      "[16:01:16] Working on T18_cluster_1_r1 ...\n",
      "[16:01:16] Working on T9_cluster_4_r2 ...\n",
      "[16:01:16] T10_cluster_1_r2.fas is not in the accepted input format, skipping\n",
      "[16:01:16] T2_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:16] T14_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:16] T30_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:16] Working on T14_cluster_1_r1 ...\n",
      "[16:01:17] Working on T30_cluster_5_r1 ...\n",
      "[16:01:17] Working on T13_cluster_1_r1 ...\n",
      "[16:01:17] T17_cluster_0_r11.fas is not in the accepted input format, skipping\n",
      "[16:01:17] T30_cluster_3_r50.fas is not in the accepted input format, skipping\n",
      "[16:01:17] T20_cluster_0_r2.fas is not in the accepted input format, skipping\n",
      "[16:01:17] Working on T30_cluster_3_r50 ...\n",
      "[16:01:34] T9_cluster_4_r2.fas is not in the accepted input format, skipping\n",
      "[16:01:34] Working on T9_cluster_3_r15 ...\n",
      "[16:01:38] T9_cluster_6_r24.fas is not in the accepted input format, skipping\n",
      "[16:01:38] Working on T14_cluster_3_r1 ...\n",
      "[16:01:38] Working on T8_cluster_0_r2 ...\n",
      "[16:01:38] T11_cluster_5_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:38] Working on T9_cluster_2_r10 ...\n",
      "[16:01:41] Working on T11_cluster_2_r1 ...\n",
      "[16:01:41] T18_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:41] T3_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:41] Working on T4_cluster_2_r1 ...\n",
      "[16:01:41] T5_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:41] T2_cluster_4_r7.fas is not in the accepted input format, skipping\n",
      "[16:01:41] T9_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:41] Working on T19_cluster_0_r1 ...\n",
      "[16:01:42] T10_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:42] Working on T6_cluster_0_r1 ...\n",
      "[16:01:42] T11_cluster_3_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:42] T21_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:42] Working on T14_cluster_4_r3 ...\n",
      "[16:01:43] T4_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:43] Working on T13_cluster_2_r1 ...\n",
      "[16:01:43] T18_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:43] Working on T9_cluster_5_r1 ...\n",
      "[16:01:43] Working on T11_cluster_5_r1 ...\n",
      "[16:01:43] T30_cluster_5_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:43] Working on T30_cluster_1_r1 ...\n",
      "[16:01:44] T7_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:44] Working on T2_cluster_0_r1 ...\n",
      "[16:01:44] T26_cluster_0_r13.fas is not in the accepted input format, skipping\n",
      "[16:01:44] Working on T3_cluster_0_r2 ...\n",
      "[16:01:45] T27_cluster_0_r5.fas is not in the accepted input format, skipping\n",
      "[16:01:45] Working on T24_cluster_1_r1 ...\n",
      "[16:01:45] T21_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:45] Working on T10_cluster_2_r1 ...\n",
      "[16:01:45] T30_cluster_4_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:45] T18_cluster_3_r1.fas is not in the accepted input format, skipping\n",
      "[16:01:45] Working on T13_cluster_0_r103 ...\n",
      "[16:02:26] T6_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:02:26] T3_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:02:26] Working on T17_cluster_0_r11 ...\n",
      "[16:02:29] T24_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:02:29] Working on T14_cluster_0_r1 ...\n",
      "[16:02:29] Working on T31_cluster_1_r2 ...\n",
      "[16:02:29] T3_cluster_0_r2.fas is not in the accepted input format, skipping\n",
      "[16:02:29] Working on T11_cluster_6_r130 ...\n",
      "[16:03:22] T14_cluster_3_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:22] Working on T11_cluster_4_r1 ...\n",
      "[16:03:23] Working on T10_cluster_1_r2 ...\n",
      "[16:03:23] T18_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:23] Working on T5_cluster_0_r1 ...\n",
      "[16:03:24] Working on T4_cluster_1_r11 ...\n",
      "[16:03:27] T14_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:27] Working on T21_cluster_1_r1 ...\n",
      "[16:03:27] Working on T9_cluster_6_r24 ...\n",
      "[16:03:36] Working on T17_cluster_1_r1 ...\n",
      "[16:03:36] Working on T26_cluster_0_r13 ...\n",
      "[16:03:39] T11_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:39] T14_cluster_4_r3.fas is not in the accepted input format, skipping\n",
      "[16:03:39] T9_cluster_5_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:39] T2_cluster_0_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:39] Working on T7_cluster_0_r1 ...\n",
      "[16:03:40] T2_cluster_3_r9.fas is not in the accepted input format, skipping\n",
      "[16:03:40] T9_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:40] Working on T18_cluster_0_r1 ...\n",
      "[16:03:40] T30_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:40] Working on T30_cluster_4_r1 ...\n",
      "[16:03:40] Working on T12_cluster_1_r1 ...\n",
      "[16:03:40] Working on T27_cluster_0_r5 ...\n",
      "[16:03:41] T2_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:41] T9_cluster_2_r10.fas is not in the accepted input format, skipping\n",
      "[16:03:41] Working on T18_cluster_3_r1 ...\n",
      "[16:03:41] Working on T2_cluster_3_r9 ...\n",
      "[16:03:43] T11_cluster_6_r130.fas is not in the accepted input format, skipping\n",
      "[16:03:43] T10_cluster_0_r7.fas is not in the accepted input format, skipping\n",
      "[16:03:43] T13_cluster_2_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:43] Working on T11_cluster_0_r1 ...\n",
      "[16:03:43] Working on T2_cluster_1_r1 ...\n",
      "[16:03:44] T17_cluster_1_r1.fas is not in the accepted input format, skipping\n",
      "[16:03:44] T12_cluster_0_r5.fas is not in the accepted input format, skipping\n",
      "[16:03:44] T12_cluster_1_r1.fas is not in the accepted input format, skipping\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'/home/raingel/桌面/nanopore_working/nanoACT/nanoact/sample/8_consensus'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#利用mafft將序列進行多序列排比，並且取得序列的consensus sequence\n",
    "#本步驟完成後，建議先檢查各個序列的排比檔(aln_)是否有異質的情形(長短不一、align不良)\n",
    "\n",
    "dumb.mafft_consensus (src=\"./nanoact/sample/7_cluster/\", \n",
    "                      des=\"./nanoact/sample/8_consensus/\",\n",
    "                      minimal_reads=0,  #minimal_reads: 設定最少的序列數量，預設為0。例如minimal_reads=10，則只會輸出序列數量大於10的consensus sequence。\n",
    "                      input_format=\"fastq\"\n",
    "                    )  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16:12:00] Blasting 0 to 30 of 69\n",
      "[16:12:02] Query D19UHBH9016 submitted.\n",
      "[16:12:02] You can check the status at https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_OBJECT=SearchInfo&RID=D19UHBH9016\n",
      "[16:12:02] And results here: https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&RID=D19UHBH9016\n",
      "[16:12:34] Search complete, retrieving results...\n",
      "[16:12:34] Retrieving results fromhttps://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&RID=D19UHBH9016&FORMAT_TYPE=XML[16:12:43] 'TaxaSet'\n",
      "[16:15:05] string indices must be integers\n",
      "[16:15:34] Blasting 30 to 60 of 69\n",
      "[16:15:37] Query D1A18A55013 submitted.\n",
      "[16:15:37] You can check the status at https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_OBJECT=SearchInfo&RID=D1A18A55013\n",
      "[16:15:37] And results here: https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&RID=D1A18A55013\n",
      "[16:16:08] Search complete, retrieving results...\n",
      "[16:16:08] Retrieving results fromhttps://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&RID=D1A18A55013&FORMAT_TYPE=XML[16:18:30] Blasting 60 to 90 of 69\n",
      "[16:18:32] Query D1A6RS6S016 submitted.\n",
      "[16:18:32] You can check the status at https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&FORMAT_OBJECT=SearchInfo&RID=D1A6RS6S016\n",
      "[16:18:32] And results here: https://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&RID=D1A6RS6S016\n",
      "[16:19:03] Search complete, retrieving results...\n",
      "[16:19:03] Retrieving results fromhttps://blast.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Get&RID=D1A6RS6S016&FORMAT_TYPE=XML[16:19:27] string indices must be integers\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'./nanoact/sample/8_consensus//blast.csv'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#將consensus序列進行blast，並生成一個csv檔，內包含每個序列的blast結果\n",
    "#Input: A folder containing fasta files named in the specified format\n",
    "#Format: con_{sampleID}_cluster_{number}_r{reads_number}\n",
    "#e.g.: con_2523_cluster_1_r499\n",
    "#Output: A csv file named {name} is saved in the {des} folder\n",
    "# `funguild`: This parameter is a boolean value that indicates whether to perform a Funguild search or not.\n",
    "#             Funguild is a web-based annotation tool that allows users to predict the ecological functions of fungal communities based on their taxonomic composition.\n",
    "# `startswith`: This parameter is a string indicating the prefix that the input fasta file names should start with. It is used to filter out files that do not match the given prefix.\n",
    "# `query_range`: 代表該序列要用來送去blast的區間。一般建議不要超過500 bp。過長除了會造成blast伺服器負擔過重外，由於blast的排序同時考慮coverage及similarity。過長\n",
    "#                的query將會導致高coverage但低similarity的hit排在前面，排擠掉中等coverage但高similarity的hit\n",
    "#                輸入值為tuple，範例： 假設序列為 AAATTTCCC\n",
    "#                query_range=(None,None)則代表完全不裁切\n",
    "#                query_range=(0,None)也代表完全不裁切\n",
    "#                query_range=(0,-1) 則代表從第0個位置(第1個 bp，程式上習慣從0開始計數)開始，到(不包含)倒數最後一個，實際query為 AAATTTCC\n",
    "#                query_range=(2,5) 則代表從第2個開始，到(不包含)第5個，實際query為 ATT\n",
    "#                query_range=(3,-3) 則代表從第3個開始，到(不包含)倒數第3個，實際query為 TTT\n",
    "# `batch`: This parameter is an integer indicating the number of sequences to blast at a time.\n",
    "#          The input sequences are divided into batches of size `batch`, and each batch is blasted separately.\n",
    "#          This is done to avoid overloading the NCBI BLAST server with too many requests at once.\n",
    "dumb.blast_2(src= \"./nanoact/sample/8_consensus/\", #Input: 一個資料夾，資料夾中包含以SampleID為檔名的fasta檔案，例如 SampleID.fasta\n",
    "           des= \"./nanoact/sample/8_consensus/\", #Output: 一個資料夾，程式會在該資料夾中輸出一個csv檔案，檔案\n",
    "           name=\"blast.csv\", #name: 輸出檔案的檔名，預設為blast.csv。\n",
    "           funguild=True, #funguild: 是否進行funguild的分析，預設為True。\n",
    "           startswith =\"con_\", #startswith: 輸入fasta檔案的檔名所具有的開頭，預設為con_。\n",
    "           input_format=\"fas\", #input_format: 輸入fasta檔案的格式，預設為fasta, 若為fas則為副檔名為fas的fasta檔案。\n",
    "           query_range=(200,400), #`query_range`: 代表該序列要用來送去blast的區間。一般建議不要超過500 bp。過長除了會造成blast伺服器負擔過重外，由於blast的排序同時考慮coverage及similarity。過長\n",
    "                                #的query將會導致高coverage但低similarity的hit排在前面，排擠掉中等coverage但高similarity的hit\n",
    "                                #輸入值為tuple，範例： 假設序列為 AAATTTCCC\n",
    "                                #query_range=(None,None)則代表完全不裁切\n",
    "                                #query_range=(0,None)也代表完全不裁切\n",
    "                                #query_range=(0,-1) 則代表從第0個位置(第1個 bp，程式上習慣從0開始計數)開始，到(不包含)倒數最後一個，實際query為 AAATTTCC\n",
    "                                #query_range=(2,5) 則代表從第2個開始，到(不包含)第5個，實際query為 ATT\n",
    "                                #query_range=(3,-3) 則代表從第3個開始，到(不包含)倒數第3個，實際query為 TTT\n",
    "           batch = 30 #`batch`: This parameter is an integer indicating the number of sequences to blast at a time.\n",
    "                      #The input sequences are divided into batches of size `batch`, and each batch is blasted separately.\n",
    "                      #This is done to avoid overloading the NCBI BLAST server with too many requests at once.\n",
    "           )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[17:51:16] Downloading custom database from NCBI...\n",
      "[17:51:18] Getting taxinfo for each record...\n",
      "[17:51:18] 6/6 taxid processed...\n",
      "[17:51:18] Merging custom database and ref_db...\n",
      "[17:51:18] Downloading ref_db: fungi.ITS[17:51:20] Downloading ref_db: bacteria.16SrRNA[17:51:25] Building ref_db from ref_db.fas...\n",
      "[17:51:54] Processing file: T2.fastq\n",
      "[17:51:54] Processing file: T6.fastq\n",
      "[17:51:55] Processing file: T4.fastq\n",
      "[17:51:56] Processing file: T14.fastq\n",
      "[17:51:56] Processing file: T20.fastq\n",
      "[17:51:57] Processing file: T18.fastq\n",
      "[17:51:57] Processing file: T3.fastq\n",
      "[17:51:58] Processing file: T25.fastq\n",
      "[17:51:58] Processing file: T7.fastq\n",
      "[17:51:58] Processing file: T24.fastq\n",
      "[17:51:59] Processing file: T27.fastq\n",
      "[17:51:59] Processing file: T29.fastq\n",
      "[17:51:59] Processing file: T11.fastq\n",
      "[17:52:00] Processing file: T21.fastq\n",
      "[17:52:00] Processing file: T30.fastq\n",
      "[17:52:00] Processing file: T31.fastq\n",
      "[17:52:01] Processing file: T5.fastq\n",
      "[17:52:01] Processing file: T28.fastq\n",
      "[17:52:01] Processing file: T13.fastq\n",
      "[17:52:02] Processing file: T12.fastq\n",
      "[17:52:02] Processing file: T1.fastq\n",
      "[17:52:02] Processing file: T9.fastq\n",
      "[17:52:03] Processing file: T23.fastq\n",
      "[17:52:03] Processing file: T16.fastq\n",
      "[17:52:03] Processing file: T19.fastq\n",
      "[17:52:03] Processing file: T22.fastq\n",
      "[17:52:03] Processing file: T15.fastq\n",
      "[17:52:03] Processing file: T10.fastq\n",
      "[17:52:04] Processing file: T17.fastq\n",
      "[17:52:04] Processing file: T8.fastq\n",
      "[17:52:05] Processing file: T26.fastq\n"
     ]
    }
   ],
   "source": [
    "#我們也可以直接將trim過的raw reads直接對refseq等資料庫進行比對\n",
    "#但這種方法由於計算量較大，所以無法直接使用blast的演算法進行\n",
    "#這裡我們使用mmseqs2的演算法進行比對\n",
    "#輸出的檔案中taxonomyResultReport可以使用pavian進行視覺化\n",
    "#https://fbreitwieser.shinyapps.io/pavian/\n",
    "#taxonomyResultReport.html是使用krona進行視覺化的結果\n",
    "#taxonomyResult.tsv是比對結果，並且可用taxonomy_assign_visualizer()進行視覺化\n",
    "dumb.taxonomy_assign(\n",
    "                        src= \"./nanoact/sample/6_trimmed/\", #Input: 一個資料夾，資料夾中包含以SampleID為檔名的fastq檔案或是fasta檔案（由input_format決定），例如 SampleID.fastq\n",
    "                        des= \"./nanoact/sample/6_1_taxonomy_assign/\", \n",
    "                        input_format='fastq', #input_format: 輸入檔案的格式，預設為fastq。可以是fastq或是fasta。\n",
    "                        lca_mode = 3, #lca_mode: 解釋較為複雜，請見mmseqs2的文件: https://mmseqs.com/latest/userguide.pdf\n",
    "                        custom_acc = ['LC729284', 'LC729293', 'LC729281', 'LC729294', 'LC729290', 'LC729267', 'LC729273'],\n",
    "                        #custom_acc為一個自訂的accession no.的列表，可以讓mmseqs2加入進行比對\n",
    "                        custom_gbff = [],\n",
    "                        #custom_gbff為一個自訂的gbff檔案的\"網址\"，可以讓mmseqs2加入進行比對\n",
    "                        ref_db = ['fungi.ITS','bacteria.16SrRNA'],#ref_db是ncbi的refdb資料庫，目前支援的資料庫如下\n",
    "                        #archaea.16SrRNA\n",
    "                        #archaea.23SrRNA\n",
    "                        #archaea.5SrRNA\n",
    "                        #bacteria.16SrRNA\n",
    "                        #bacteria.23SrRNA\n",
    "                        #bacteria.5SrRNA\n",
    "                        #fungi.18SrRNA\n",
    "                        #fungi.28SrRNA\n",
    "                        #fungi.ITS\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#custom_taxonomy_sankey是用來將mmseqs2輸出的taxonomyResult.tsv進行Sankey diagram的視覺化\n",
    "dumb.custom_taxonomy_sankey(src=\"./nanoact/sample/6_1_taxonomy_assign/\",  #Input: 一個資料夾，資料夾中包含以SampleID為檔名的taxonomyResult.tsv檔案，例如 SampleID.taxonomyResult.tsv\n",
    "                                des=\"./nanoact/sample/6_1_taxonomy_assign/\",  #Output: 一個資料夾，程式會在該資料夾中輸出以SampleID為檔名的檔案\n",
    "                                img_ext=\"png\", #img_ext: 輸出檔案的格式，預設為png。可以matplotlib支援的格式，如png、svg、jpg等。\n",
    "                                minimal_reads=2, #當一個taxon的reads數量小於minimal_reads時則不顯示\n",
    "                                vertical_scale=0.8 #圖片的垂直高度，需要視情況調整，若發現輸出的圖片有字重疊的情形，可以把vertical_scale調大\n",
    "                                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
